### PR TITLE
support for maintenance start time configuration (15331)

### LINKED ISSUE #15331: aws_storagegateway_gateway: support for maintenance window configuration
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently aws_storagegateway_gateway does not support maintenance window configuration, while AWS API exist for that.

### Potential Terraform Configuration

```hcl
resource "aws_storagegateway_gateway" "storage_gateway" {
  gateway_name             = "storage-gateway"
  maintenance_window {
    hour = 22
    minute = 0
    dow = 5
  }
  ....
}
```

### References

Same functionality in the AWS CLI: https://docs.aws.amazon.com/cli/latest/reference/storagegateway/update-maintenance-start-time.html
--->

* #0000

### PR DESCRIPTION
**Pull request by awasilyev**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #15331

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
aws_storagegateway_gateway: configurable maintenance start time
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSStorageGatewayGateway_maintenanceStartTime'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSStorageGatewayGateway_maintenanceStartTime -timeout 120m
=== RUN   TestAccAWSStorageGatewayGateway_maintenanceStartTime
=== PAUSE TestAccAWSStorageGatewayGateway_maintenanceStartTime
=== CONT  TestAccAWSStorageGatewayGateway_maintenanceStartTime
--- PASS: TestAccAWSStorageGatewayGateway_maintenanceStartTime (223.15s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	225.609s
```

### CODE REVIEW

File: aws/resource_aws_storagegateway_gateway.go [hunk 1]
```diff
@@ -156,6 +156,32 @@ func resourceAwsStorageGatewayGateway() *schema.Resource {
 				Optional:     true,
 				ValidateFunc: validation.IntAtLeast(51200),
 			},
+			"maintenance_start_time": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"hour": {
```
Comment (DrFaust92) [inline aws/resource_aws_storagegateway_gateway.go:165]: Lets keep parity with api naming: `day_of_week`, `hour_of_day` etc.

File: aws/resource_aws_storagegateway_gateway.go [hunk 2]
```diff
@@ -311,6 +337,23 @@ func resourceAwsStorageGatewayGatewayCreate(d *schema.ResourceData, meta interfa
 		}
 	}
 
+	if v, ok := d.GetOk("maintenance_start_time"); ok && len(v.([]interface{})) > 0 {
+		m := v.([]interface{})[0].(map[string]interface{})
+
+		input := &storagegateway.UpdateMaintenanceStartTimeInput{
```
Comment (DrFaust92) [inline aws/resource_aws_storagegateway_gateway.go:343]: lets also add DayOfMonth here while you are at it.

File: aws/resource_aws_storagegateway_gateway_test.go [hunk 3]
```diff
@@ -1106,3 +1145,19 @@ resource "aws_storagegateway_gateway" "test" {
 }
 `, rName, rate)
 }
+
+func testAccAWSStorageGatewayGatewayMaintenanceStartTimeConfig(rName string, hour int, minute int, day_of_week int) string {
+	return testAccAWSStorageGateway_TapeAndVolumeGatewayBase(rName) + fmt.Sprintf(`
+resource "aws_storagegateway_gateway" "test" {
```
Comment (DrFaust92) [inline aws/resource_aws_storagegateway_gateway_test.go:1151]: format test config

**Review by DrFaust92 — CHANGES_REQUESTED** [review]
some minor changes and ill verify the tests

**Comment by DrFaust92** [comment]
Also might want to add computed to the added arguments as i believe there is a default start time when none is supplied. there may be a config drift otherwise. 

also add check for basic case to fetch these computed values when `maintenance_start_time` is not supllied.

**Review by DrFaust92 — CHANGES_REQUESTED** [review]
Also might want to add computed to the added arguments as i believe there is a default start time when none is supplied. there may be a config drift otherwise.

also add check for basic case to fetch these computed values when maintenance_start_time is not supplied.

File: aws/resource_aws_storagegateway_gateway.go (hunk 1 above)
Comment (awasilyev) [reply aws/resource_aws_storagegateway_gateway.go]: done

File: aws/resource_aws_storagegateway_gateway_test.go (hunk 3 above)
Comment (awasilyev) [reply aws/resource_aws_storagegateway_gateway_test.go:1220]: done

File: aws/resource_aws_storagegateway_gateway.go (hunk 2 above)
Comment (awasilyev) [reply aws/resource_aws_storagegateway_gateway.go:412]: done

**Comment by awasilyev** [comment]
@DrFaust92 it is failing for some reason on Acceptance Test Linting, while locally terrafmt is running fine:

```
~/terraform-provider-aws (15331)$ ~/go/bin/terrafmt diff --check --fmtcompat aws/resource_aws_storagegateway_gateway_test.go 
~/terraform-provider-aws (15331)$ echo $?
0
```

**Comment by cjbeckwith** [comment]
As of May 27th I don't see that this has been implemented (at least as far as the documentation is concerned).  Has the maintenance_windows parameter been added to the resource yet?  Any ETA?

**Comment by ewbankkit** [comment]
@awasilyev Thanks for the contribution :tada: :clap:.
This change has been rolled into #24038.

**Pull request closed without merging**